pkg/agent/controller/servicecidr/provider_test.go

```diff
@@ -0,0 +1,117 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package servicecidr
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+type fakeProvider struct {
+	Interface
+	client          *fake.Clientset
+	informerFactory informers.SharedInformerFactory
+}
+
+func newFakeProvider(objects []runtime.Object) *fakeProvider {
+	client := fake.NewSimpleClientset(objects...)
+	informerFactory := informers.NewSharedInformerFactory(client, 0)
+	svcInformer := informerFactory.Core().V1().Services()
+
+	return &fakeProvider{
+		Interface:       NewServiceCIDRProvider(svcInformer),
+		client:          client,
+		informerFactory: informerFactory,
+	}
+}
+
+func makeService(namespace, name string, clusterIP string, protocol corev1.Protocol) *corev1.Service {
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Annotations: map[string]string{},
+		},
+		Spec: corev1.ServiceSpec{
+			ClusterIPs: []string{clusterIP},
+			ClusterIP:  clusterIP,
+			Ports: []corev1.ServicePort{
+				{
+					Name:     "80",
+					Port:     int32(80),
+					Protocol: protocol,
+				},
+			},
+			Type: corev1.ServiceTypeClusterIP,
+		},
+	}
+	return svc
+}
+
+func TestServiceCIDRProvider(t *testing.T) {
+	svc1 := makeService("ns1", "svc1", "10.10.0.1", corev1.ProtocolTCP)
+	svc2 := makeService("ns1", "svc2", "10::1", corev1.ProtocolTCP)
+	fc := newFakeProvider([]runtime.Object{svc1, svc2})
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	fc.informerFactory.Start(stopCh)
+	fc.informerFactory.WaitForCacheSync(stopCh)
+
+	svc3 := makeService("ns1", "svc3", "10.10.0.2", corev1.ProtocolTCP)
+	_, err := fc.client.CoreV1().Services("ns1").Create(context.TODO(), svc3, metav1.CreateOptions{})
+	assert.NoError(t, err)
+	assert.NoError(t, checkClusterIPCIDR("10.10.0.0/30", fc, false))
+
+	svc4 := makeService("ns1", "svc4", "10.10.0.5", corev1.ProtocolTCP)
+	_, err = fc.client.CoreV1().Services("ns1").Create(context.TODO(), svc4, metav1.CreateOptions{})
+	assert.NoError(t, err)
+	assert.NoError(t, checkClusterIPCIDR("10.10.0.0/29", fc, false))
+
+	svc5 := makeService("ns1", "svc5", "10::1", corev1.ProtocolTCP)
+	_, err = fc.client.CoreV1().Services("ns1").Create(context.TODO(), svc5, metav1.CreateOptions{})
+	assert.NoError(t, err)
+	assert.NoError(t, checkClusterIPCIDR("10::1/128", fc, true))
+
+	svc6 := makeService("ns1", "svc6", "10::ffff", corev1.ProtocolTCP)
+	_, err = fc.client.CoreV1().Services("ns1").Create(context.TODO(), svc6, metav1.CreateOptions{})
+	assert.NoError(t, err)
+	assert.NoError(t, checkClusterIPCIDR("10::/112", fc, true))
```

Comment:
It should also test the event handler, and perhaps leverage it to avoid polling.